materialize-snowflake/snowflake.go

```diff
@@ -430,93 +442,160 @@ func (d *transactor) Load(it *pm.LoadIterator, loaded func(int, json.RawMessage)
+		if err != nil {
+			return nil, err
+		}
+
+		if b.store.mustMerge {
+			d.cp[b.target.StateKey] = &checkpointItem{
+				Table:    b.target.Identifier,
+				Query:    renderWithDir(b.store.mergeInto, dir),
+				ToDelete: dir,
+			}
+		} else {
+			d.cp[b.target.StateKey] = &checkpointItem{
+				Table:    b.target.Identifier,
+				Query:    renderWithDir(b.store.copyInto, dir),
+				ToDelete: dir,
+			}
 		}
 	}
 
-	return func(ctx context.Context, runtimeCheckpoint *protocol.Checkpoint, _ <-chan struct{}) (*pf.ConnectorState, pf.OpFuture) {
-		log.Info("store: starting commit phase")
-		var err error
-		if d.store.fence.Checkpoint, err = runtimeCheckpoint.Marshal(); err != nil {
-			return nil, pf.FinishedOperation(fmt.Errorf("marshalling checkpoint: %w", err))
+	return func(ctx context.Context, runtimeCheckpoint *protocol.Checkpoint) (*pf.ConnectorState, m.OpFuture) {
+		var checkpointJSON, err = json.Marshal(d.cp)
+		if err != nil {
+			return nil, m.FinishedOperation(fmt.Errorf("creating checkpoint json: %w", err))
 		}
 
-		return nil, sql.CommitWithDelay(ctx, d.store.round, d.updateDelay, it.Total, d.commit)
+		return &pf.ConnectorState{UpdatedJson: checkpointJSON, MergePatch: true}, nil
 	}, nil
 }
 
-func (d *transactor) commit(ctx context.Context) error {
-	var txn, err = d.store.conn.BeginTx(ctx, nil)
-	if err != nil {
-		return fmt.Errorf("store.conn.BeginTx: %w", err)
-	}
-	defer txn.Rollback()
-
-	// First we must validate the fence has not been modified.
-	var fenceUpdate strings.Builder
-	if err := d.templates["updateFence"].Execute(&fenceUpdate, d.store.fence); err != nil {
-		return fmt.Errorf("evaluating fence template: %w", err)
-	} else if _, err = txn.ExecContext(ctx, fenceUpdate.String()); err != nil {
-		err = fmt.Errorf("txn.Exec: %w", err)
+func renderWithDir(tpl string, dir string) string {
+	return fmt.Sprintf(tpl, dir)
+}
 
-		// Give recommendation to user for resolving timeout issues
-		if strings.Contains(err.Error(), "timeout") {
-			return fmt.Errorf("fence.Update: %w  (ensure LOCK_TIMEOUT and STATEMENT_TIMEOUT_IN_SECONDS are at least ten minutes)", err)
+// Acknowledge merges data from temporary table to main table
+func (d *transactor) Acknowledge(ctx context.Context) (*pf.ConnectorState, error) {
+	var asyncCtx = sf.WithAsyncMode(ctx)
+	log.Info("store: starting committing changes")
+
+	// Run the queries using AsyncMode, which means that `ExecContext` will not block
+	// until the query is successful, rather we will store the results of these queries
+	// in a map so that we can then call `RowsAffected` on them, blocking until
+	// the queries are actually executed and done
+	var results = make(map[string]stdsql.Result)
+	for stateKey, item := range d.cp {
+		if len(item.Query) == 0 {
+			continue
+		}
+		// during recovery we skip queries that belong to tables which do not have a binding anymore
+		// since these tables might be deleted already
+		if !d.hasTableBinding(item.Table) {
```

Comment:
I think we need to be looking for if the spec has the `stateKey` included or not. Otherwise, when re-backfilling, we may have multiple `stateKey`'s for the same table, and would potentially encounter persisted queries for the same table multiple times.